content/post/71-putting-npm-on-ipfs-part-2.md

```diff
@@ -0,0 +1,82 @@
+
+`ipfs-npm` proxies for the npm cli, passing all arguments on, so use it as you would npm, just with a different command name:
+
+```console
+$ cd my-project
+$ ipfs-npm install
+👿 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
+Swarm listening on /ip4/127.0.0.1/tcp/57029/ipfs/QmZ7vEvXRdTVipb9o2p2Cmt3s1S8rqbo2ohscTjvTLgnpP
+🗂️ Loading registry index from https://registry.js.ipfs.io
+☎️ Dialling registry mirror /ip4/35.178.192.119/tcp/10015/ipfs/QmWBaYSnmgZi6F6D69JuZGhyL8rm6pt8GX5r7Atc6Gd7vR,/dns4/registry.js.ipfs.io/tcp/10015/ipfs/QmWBaYSnmgZi6F6D69JuZGhyL8rm6pt8GX5r7Atc6Gd7vR
+📱️ Connected to registry
+👩‍🚀 Starting local proxy
+🚀 Server running on port 37847
+🎁 Installing dependencies with /Users/alex/.nvm/versions/node/v10.15.1/bin/npm
+...
+```
+
+If you prefer Yarn and have it installed globally, use the `ipfs-yarn` command (installed alongside `ipfs-npm`):
+
+```console
+$ cd my-project
+$ ipfs-yarn
+👿 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
+Swarm listening on /ip4/127.0.0.1/tcp/49905/ipfs/QmZ7vEvXRdTVipb9o2p2Cmt3s1S8rqbo2ohscTjvTLgnpP
+🗂️  Loading registry index from https://registry.js.ipfs.io
+☎️  Dialling registry mirror /ip4/35.178.192.119/tcp/10040/ipfs/QmfKqxieE71QoAchNk5e2MKmvWKjGdUnSifHqq1xZLEzyn,/dns4/registry.js.ipfs.io/tcp/10040/ipfs/QmfKqxieE71QoAchNk5e2MKmvWKjGdUnSifHqq1xZLEzyn
+📱️ Connected to registry
+👩‍🚀 Starting local proxy
+🚀 Server running on port 49935
+🎁 Installing dependencies with /Users/alex/.nvm/versions/node/v10.15.1/bin/yarn
+yarn install v1.13.0
+...
+```
+
+## 🙋 How does it work?
+
+Behind the scenes, it starts a local http server and configures npm/Yarn to use it, it then uses the same package metadata as [registry.js.ipfs.io](https://registry.js.ipfs.io) to resolve a CID for a requested tarball - if it has the CID, it uses IPFS to request the content, otherwise it downloads it from the public npm registry, adds it to IPFS and stores the CID for the next time it's requested.
+
+![Request sequence](/71-putting-npm-on-ipfs-part-2/ipfs-npm-sequence.png)
+
+If this sounds familiar it may be because you've just read [part 1](/post/70-putting-npm-on-ipfs-part-1) - indeed it's the exact same code fulfilling the exact same function, just locally on your machine instead of on a server somewhere.
+
+The http server is necessary because at the time of writing neither npm or Yarn support IPFS as a transport - hopefully one day this won't be necessary but as it stands no http traffic leaves your machine, unless you request modules that don't have CIDs in their metadata.
+
+It also dials the ipfs-npm-registry-mirror directly to improve the speed of content resolution - this is partly because we know that the mirror will have the content we are after but also because there is no [DHT](https://ipfs.io/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/wiki/Distributed_hash_table.html) in js-IPFS (yet - it's coming!) so discovery would have to occur via the gateway nodes.
```

Comment:
isn't DHT available as of v0.35.0, or am I misreading the [release notes](https://github.com/ipfs/js-ipfs/issues/1826)?

it might be worth mentioning that in the future this is something that could be handled by rendezvous in the future, as well

Reply:
@parkan I don't think v0.35.0 final has been published to npm yet

Reply:
Quite right.  It may not be enabled by default until v0.36.0 either.